### ISSUE TITLE
Image_picker couldn't be closed by dragging it down

### ISSUE THREAD

**Issue by marinalvdgs** [flutter/flutter#87330]
Here code sample:
```Dart
pickedFile = await picker.pickImage(source: ImageSource.gallery);
if (pickedFile != null) {
      fileUrl = await saveFileToApplication(
          File(pickedFile.path), getUniqPhotoName());
    }
```
When I try to close the dialog by tapping the cancel button it works normally. But when I try to close it by sliding down, it closes only visually without returning anything.
## Steps to Reproduce

1. Open image picker with gallery source 
2. Try to close it by dragging it down

**Expected results:** Picker closed, the app works normally

**Actual results:** Picker doesn't really close, app freezes

### More Information
I can reproduce this issue on iOS 14.3 and higher.
Also, I was faced with the same issue using another package for picking files. If it will be helpful, [here is the solution](https://github.com/kineapps/flutter_file_dialog/commit/9b644d17af6a1734868528d52adbe80948f932a5) by which developers fixed this bug.
<details>
  <summary>Flutter doctor logs</summary>

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-x64, locale en-RU)
    • Flutter version 2.2.3 at /Users/marinaivanova/dev/flutter
    • Framework revision f4abaa0735 (4 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/marinaivanova/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] Connected device (2 available)
    • iPhone 12 Pro (mobile) • 9086F30F-4A99-48AE-855C-2DE615267DC9 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-5
      (simulator)
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 92.0.4515.107

• No issues found!
```

</details>

**Comment by maheshmnj** [comment]
Hi @marinalvdgs, May I know what version of the `image_picker` plugin are you using? I tried using the latest `image_picker: ^0.8.2` plugin and I was able to slide down to close the selection.

<details>
<summary>Output</summary>

https://user-images.githubusercontent.com/31410839/127646442-8e544528-081f-4b63-9922-de8b3b512679.mov

</details>


<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-arm, locale en-GB)
    • Flutter version 2.2.3 at /Users/mahesh/Documents/flutter
    • Framework revision f4abaa0735 (4 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 57.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (4 available)
    • Redmi K20 Pro (mobile)     • 192.168.1.3:5555                     • android-arm64  • Android 11 (API 30)
    • iPhone 12 Pro Max (mobile) • 6A434744-B348-4FB3-AB10-131292174456 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-4 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 92.0.4515.107

• No issues found!
```

</details>

**Comment by marinalvdgs** [comment]
Hi @maheshmnj, I use `image_picker: ^0.8.2` too. In my code sample when I slide down the picker, looks like it doesn't return any value.
I've tried to add a breakpoint on the `if (pickedFile != null)` line, but debugger never achieves it. 
In my case, I block UI with loader while image dialog doesn't return anything, and after sliding down UI is still blocked.
I think this behavior is a bug.
Thanks for the response!

**Comment by maheshmnj** [comment]
Could you please share the minimal reproducible code and the logs to better address the issue.

**Comment by marinalvdgs** [comment]
Ok, here is the full example:
```Dart
Future<String?> getAndSaveImage(){
  XFile? pickedFile;
  try {
      pickedFile = await picker.pickImage(source: ImageSource.gallery);
    } on PlatformException catch (e, stackTrace) {
      print('$e $stackTrace');
   }
  if (pickedFile != null) {
      //save file
      print(pickedFile.path);
    }
  return pickedFile?.path ?? null;
}

void openGallery() async {
  bool isDialogClosed = false;
  showDialog(
    barrierDismissible: false,
    context: context,
    builder: (BuildContext context) {
      return WillPopScope(
        child: Center(
          child: CircularProgressIndicator(),
        ),
        onWillPop: () {
          return Future.value(false);
        },
      );
    },
  ).then((value) {
    isDialogClosed = true;
  });
  try {
    await getAndSaveImage();
  } catch (e, stackTrace) {
    print(stackTrace);
    showSnackbar(context, title: e.toString());
  } finally {
    if (!isDialogClosed &&
        Navigator.of(context, rootNavigator: true).canPop()) {
      Navigator.of(context, rootNavigator: true).pop();
    }
  }
}
```

It doesn't throw any exceptions or something so my log output is empty. But I can show you how it looks. First time I close the picker by tapping the cancel button and it works normally. The second time I slide it down and see the endless loader. 

<details>
  <summary>Screen record</summary>
  

https://user-images.githubusercontent.com/55170926/127655552-993320a6-1f73-4b96-8726-1a1a17762e1d.mp4


</details>

**Comment by maheshmnj** [comment]
@marinalvdgs Thanks for the details, I am able to reproduce the issue. 
When we cancel the image picker by sliding down, the image picker should return "null" as it happens for the case when we click the cancel button.  
And I see an exception being thrown when picking the image for the second time which is basically the previous exception.

```
flutter: PlatformException(multiple_request, Cancelled by a second request, null, null)
```

<details>
<summary>code sample</summary>

```dart

import 'dart:io';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(primarySwatch: Colors.blue),
        home: ImagePickerSample()
        );
  }
}

class ImagePickerSample extends StatefulWidget {
  ImagePickerSample({Key? key, this.title = 'Image Picker Sample'})
      : super(key: key);

  final String title;

  @override
  _ImagePickerSampleState createState() => _ImagePickerSampleState();
}

class _ImagePickerSampleState extends State<ImagePickerSample> {
  Widget imageBuilder(File file, int index) {
    if (file != null && file.path.isNotEmpty) {
      return Image.file(
        File(
          file.path,
        ),
        fit: BoxFit.cover,
      );
    } else {
      return imagePlaceHolder(index);
    }
  }

  Future<XFile?> pickImage(ImageSource source) async {
    try {
      final image = await _picker.pickImage(
        source: source,
      );
      return image;
    } catch (e, stackTrace) {
      return XFile('');
    }
  }

  Widget imagePlaceHolder(int index) {
    return Container(
      // padding: EdgeInsets.only(right),
      alignment: Alignment.center,
      decoration: BoxDecoration(
        border: Border(
            right: index % 2 == 0 ? BorderSide() : BorderSide.none,
            bottom: BorderSide()),
      ),
      child: Icon(Icons.add_a_photo),
    );
  }

  final _picker = ImagePicker();

  List<File> images = [];

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    List.generate(4, (x) => images.add(File('')));
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Expanded(
              child: GridView.builder(
                itemCount: 4,
                gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
                    crossAxisCount: 2),
                itemBuilder: (context, index) => GestureDetector(
                    onTap: () async {
                      final image = await pickImage(ImageSource.gallery);
                      setState(() {
                        images[index] = File(image!.path);
                      });
                    },
                    child: imageBuilder(images[index], index)),
              ),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {},
        tooltip: 'Increment',
        child: Icon(Icons.add),
      ),
    );
  }
}

```
</details>

<details>
<summary>Output</summary>

https://user-images.githubusercontent.com/31410839/127857605-e02f7ca3-c4c6-42aa-b10a-04ad5bfd605d.mov

</details>

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-arm, locale en-GB)
    • Flutter version 2.2.3 at /Users/mahesh/Documents/flutter
    • Framework revision f4abaa0735 (4 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 57.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 91.0.4472.164

• No issues found!
```

```
[✓] Flutter (Channel master, 2.5.0-2.0.pre.69, on macOS 11.4 20F71 darwin-arm, locale en-GB)
    • Flutter version 2.5.0-2.0.pre.69 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision ccf6f0353a (10 minutes ago), 2021-07-29 01:50:54 -0700
    • Engine revision 40451453ab
    • Dart version 2.14.0 (build 2.14.0-361.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] IntelliJ IDEA Community Edition (version 2021.1.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 57.0.5
    • Dart plugin version 211.7233

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.24.0

[✓] Connected device (4 available)
    • sdk gphone arm64 (mobile)  • emulator-5554                        • android-arm64  • Android 11 (API 30) (emulator)
    • iPhone 12 Pro Max (mobile) • 6A434744-B348-4FB3-AB10-131292174456 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-4
      (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 92.0.4515.107

• No issues found!
```
</details>

**Comment by kristijorgji** [comment]
In ios emulator after I call image picker, I get error right away message 
```
PlatformException(multiple_request, Cancelled by a second request, null, null)
```
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:597:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
<asynchronous suspension>
#2      MethodChannelImagePicker.pickImage (package:image_picker_platform_interface/src/method_channel/method_channel_image_picker.dart:29:20)
```

And looks like:
https://user-images.githubusercontent.com/15630854/126806572-574f7e9c-94f3-4879-8511-acf2deea1f5a.png


Full report

https://github.com/flutter/flutter/issues/86942

**Comment by BeMacized** [comment]
I have been able to reproduce this issue as well and have been looking into it for a bit. I'm however getting stuck and I don't have much experience with iOS development. 

To me it looks like the issue is that the newer [PHPickerViewController](https://developer.apple.com/documentation/photokit/phpickerviewcontroller?language=objc) only seems to to call [didFinishPicking](https://developer.apple.com/documentation/photokit/phpickerviewcontrollerdelegate/3606609-picker?language=objc) when images have been picked, or when the user explicitly cancels the operation by pressing the cancel button. Not when cancelling by swiping away the modal. Unlike the older [UIImagePickerControllerDelegate](https://developer.apple.com/documentation/uikit/uiimagepickercontrollerdelegate), the [PHPickerViewControllerDelegate](https://developer.apple.com/documentation/uikit/uiimagepickercontrollerdelegate?language=objc) does not seem to have a separate method for detecting cancellation.

As [PHPickerViewController](https://developer.apple.com/documentation/photokit/phpickerviewcontroller?language=objc) inherits from [UIViewController](https://developer.apple.com/documentation/uikit/uiviewcontroller?language=objc), I was hoping we could maybe extend it to override [viewDidDisappear](https://developer.apple.com/documentation/uikit/uiviewcontroller/1621423-viewdidappear?language=objc) and try detect it this way, but the docs explicitly mention _"You can’t subclass PHPickerViewController because its view hierarchy is private and not accessible through the public API"_. 

Otherwise, I'm not sure how to detect [PHPickerViewController](https://developer.apple.com/documentation/photokit/phpickerviewcontroller?language=objc) being closed this way. If anyone has an idea, any help would be appreciated.

**Comment by maheshmnj** [comment]
@kristijorgji, The issue that you are reporting is filed here #82602

**Comment by the-recondite** [comment]
Any update on a fix for this?🥺. Would really appreciate that

**Comment by kristijorgji** [comment]
@maheshmnj you are linking me in cycles ;) check the other thread, you say to come and post here, on this thread you redirect me to the other one.

I do not see any fix in progress for either issue
https://github.com/flutter/flutter/issues/87330
or
https://github.com/flutter/flutter/issues/70436
or
https://github.com/flutter/flutter/issues/82602

Can you please just post here the solution/fix ?

**Comment by maheshmnj** [comment]
@kristijorgji my only point was that the exception logs that you [posted above](https://github.com/flutter/flutter/issues/87330#issuecomment-895127623) are not relevant to this issue but the linked issue 

> you say to come and post here,

Could you point me where did I say this?

**Comment by BeMacized** [comment]
A fix for this has been merged in flutter/plugins#4228. Closing.
